sdk/cosmos/azure-cosmos/azure/cosmos/_global_endpoint_manager.py

```diff
@@ -118,61 +118,102 @@ def _refresh_endpoint_list_private(self, database_account=None, **kwargs):
-        all_endpoints = [self.location_cache.read_regional_endpoints[0]]
-        all_endpoints.extend(self.location_cache.write_regional_endpoints)
-        count = 0
-        for endpoint in all_endpoints:
-            count += 1
-            if count > 3:
+        endpoints_attempted = set()
+        database_account, endpoint = self._GetDatabaseAccount(**kwargs)
+        endpoints_attempted.add(endpoint)
+        self.location_cache.perform_on_database_account_read(database_account)
+
+        # should use the endpoints in the order returned from gateway and only the ones specified in preferred locations
+        read_account_dual_endpoints_iterator = iter(self.location_cache.account_read_dual_endpoints_by_location
+                                                    .values())
+        first_read_dual_endpoint = None
+        # find first read endpoint from gateway and in preferred locations
+        for read_dual_endpoint in read_account_dual_endpoints_iterator:
+            if read_dual_endpoint in self.location_cache.read_dual_endpoints:
+                first_read_dual_endpoint = read_dual_endpoint
                 break
-            try:
-                self.Client._GetDatabaseAccountCheck(endpoint.get_current(), **kwargs)
-            except (exceptions.CosmosHttpResponseError, AzureError):
-                if endpoint in self.location_cache.read_regional_endpoints:
-                    self.mark_endpoint_unavailable_for_read(endpoint.get_current(), False)
-                if endpoint in self.location_cache.write_regional_endpoints:
-                    self.mark_endpoint_unavailable_for_write(endpoint.get_current(), False)
-                    endpoint.swap()
+        if first_read_dual_endpoint:
+            dual_endpoints = [first_read_dual_endpoint]
+        else:
+            dual_endpoints = []
+        # add write endpoints from gateway and in preferred locations
+        write_dual_endpoints = [endpoint for endpoint in
+                                self.location_cache.account_write_dual_endpoints_by_location.values()
+                                if endpoint in self.location_cache.write_dual_endpoints]
+        dual_endpoints.extend(write_dual_endpoints)
+        success_count = 0
+        for dual_endpoint in dual_endpoints:
+            if dual_endpoint.get_primary() not in endpoints_attempted:
+                if success_count >= 2:
+                    break
+                endpoints_attempted.add(dual_endpoint.get_primary())
+                try:
+                    if (dual_endpoint.get_primary() in
+                            self.location_cache.location_unavailability_info_by_endpoint):
+                        # if the endpoint is unavailable, we need to lower the timeouts to be more aggressive in the
+                        # health check. This helps reduce the time the health check is blocking all requests.
+                        previous_dba_read_timeout = self.Client.connection_policy.DBAReadTimeout
```

Comment:
i think this is not restored when the health check fails? Setting back to previous would need to happen in finally block or except as well? rpobably worth putting this in helper methods on connection_policy - like override_health_check_timeout(duration) and reset_health_check_timeoput or similar